File: app/src/main/java/com/example/myapplication/feature/today/ExerciseCard.kt
package com.example.myapplication.feature.today

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Checkbox
import androidx.compose.material3.CheckboxDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.myapplication.core.model.MuscleGroup
import com.example.myapplication.ui.theme.*

@Composable
fun ExerciseCard(
    sessionId: Long,
    row: WorkoutRowUi,
    enabled: Boolean,
    onCheckedChange: (Boolean) -> Unit,
) {
    var expanded by rememberSaveable(sessionId, row.orderIndex, row.exerciseId) { mutableStateOf(false) }

    val colors = MaterialTheme.colorScheme
    val customColors = colors.customColors

    val borderColor by animateColorAsState(
        targetValue = if (row.checked) customColors.checkedCardBorder else colors.outline,
        label = "cardBorder",
    )
    val bgColor by animateColorAsState(
        targetValue = if (row.checked) customColors.greenLight else colors.surfaceVariant,
        label = "cardBg",
    )

    Column(
        Modifier
            .fillMaxWidth()
            .animateContentSize()
            .background(bgColor, RoundedCornerShape(16.dp))
            .border(1.dp, borderColor, RoundedCornerShape(16.dp))
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            // Exercise number badge
            Box(
                modifier = Modifier
                    .size(36.dp)
                    .clip(CircleShape)
                    .background(if (row.checked) SuccessGreen else EnergyOrange),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    if (row.checked) "✓" else "${row.orderIndex + 1}",
                    color = Color.White,
                    fontWeight = FontWeight.Bold,
                    fontSize = 14.sp,
                )
            }
            Spacer(Modifier.width(12.dp))

            Column(Modifier.weight(1f)) {
                Text(
                    row.nameVi,
                    style = MaterialTheme.typography.titleMedium,
                    color = customColors.primaryText,
                    fontWeight = FontWeight.SemiBold,
                    textDecoration = if (row.checked) TextDecoration.LineThrough else TextDecoration.None,
                )
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                ) {
                    Text(
                        muscleEmoji(row.primaryMuscle),
                        fontSize = 14.sp,
                    )
                    Text(
                        "${row.prescriptionText} · nghỉ ${row.restSeconds}s",
                        color = customColors.mutedText,
                        style = MaterialTheme.typography.bodySmall,
                    )
                }
            }

            Checkbox(
                checked = row.checked,
                onCheckedChange = onCheckedChange,
                enabled = enabled,
                colors = CheckboxDefaults.colors(
                    checkedColor = SuccessGreen,
                    uncheckedColor = colors.outline,
                ),
                modifier = Modifier
                    .testTag("exercise-checkbox-${row.orderIndex}")
                    .semantics { contentDescription = "Đánh dấu ${row.nameVi} hoàn thành" },
            )
        }

        // Expandable instructions
        Text(
            if (expanded) "Ẩn hướng dẫn ▲" else "Xem hướng dẫn ▼",
            color = EnergyOrange,
            style = MaterialTheme.typography.labelMedium,
            fontWeight = FontWeight.SemiBold,
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = 48.dp)
                .testTag("exercise-expand-${row.orderIndex}")
                .clickable(enabled = enabled) { expanded = !expanded }
                .semantics {
                    contentDescription = if (expanded) "Đóng hướng dẫn ${row.nameVi}" else "Mở hướng dẫn ${row.nameVi}"
                }
                .padding(vertical = 6.dp),
        )

        if (expanded) {
            Column(
                Modifier
                    .fillMaxWidth()
                    .background(colors.surface, RoundedCornerShape(12.dp))
                    .padding(14.dp)
                    .testTag("exercise-instructions-${row.orderIndex}"),
                verticalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                row.instructionsVi.forEachIndexed { index, instruction ->
                    Row {
                        Text(
                            "${index + 1}.",
                            color = EnergyOrange,
                            fontWeight = FontWeight.Bold,
                            modifier = Modifier.width(24.dp),
                        )
                        Text(instruction, color = customColors.primaryText, style = MaterialTheme.typography.bodySmall)
                    }
                }
            }
        }
    }
}

private fun muscleEmoji(muscle: MuscleGroup): String = when (muscle) {
    MuscleGroup.CHEST -> "🫁"
    MuscleGroup.BACK -> "🔙"
    MuscleGroup.SHOULDERS -> "💪"
    MuscleGroup.BICEPS -> "💪"
    MuscleGroup.TRICEPS -> "💪"
    MuscleGroup.CORE -> "🎯"
    MuscleGroup.QUADS -> "🦵"
    MuscleGroup.HAMSTRINGS -> "🦵"
    MuscleGroup.GLUTES -> "🍑"
    MuscleGroup.CALVES -> "🦶"
    MuscleGroup.FULL_BODY -> "🏋️"
    MuscleGroup.CARDIO -> "❤️"
    MuscleGroup.MOBILITY -> "🧘"
}

File: flutter/lib/feature/today/widgets/exercise_card.dart
import 'package:flutter/material.dart';
import '../../../core/model/catalog_models.dart';
import '../../../ui/components/exercise_3d_dialog.dart';
import '../today_ui_state.dart';
import '../../../ui/theme/colors.dart';
import '../../../ui/theme/theme.dart';
import '../../../ui/theme/spacing.dart';
import '../../../ui/theme/radius.dart';
import '../../../ui/theme/typography.dart';
import '../../../ui/components/gym_card.dart';
import '../../../ui/components/gym_checkbox.dart';

class ExerciseCard extends StatefulWidget {
  final int sessionId;
  final WorkoutRowUi row;
  final bool enabled;
  final ValueChanged<bool> onCheckedChange;
  final VoidCallback onSubstitute;

  const ExerciseCard({
    super.key,
    required this.sessionId,
    required this.row,
    required this.enabled,
    required this.onCheckedChange,
    required this.onSubstitute,
  });

  @override
  State<ExerciseCard> createState() => _ExerciseCardState();
}

class _ExerciseCardState extends State<ExerciseCard> {
  bool _expanded = false;

  String _muscleEmoji(MuscleGroup muscle) {
    switch (muscle) {
      case MuscleGroup.chest:
        return "🫁";
      case MuscleGroup.back:
        return "🔙";
      case MuscleGroup.shoulders:
      case MuscleGroup.biceps:
      case MuscleGroup.triceps:
        return "💪";
      case MuscleGroup.core:
        return "🎯";
      case MuscleGroup.quads:
      case MuscleGroup.hamstrings:
        return "🦵";
      case MuscleGroup.glutes:
        return "🍑";
      case MuscleGroup.calves:
        return "🦶";
      case MuscleGroup.fullBody:
        return "🏋️";
      case MuscleGroup.cardio:
        return "❤️";
      case MuscleGroup.mobility:
        return "🧘";
    }
  }

  @override
  Widget build(BuildContext context) {
    final customColors = context.customColors;
    final isDark = Theme.of(context).brightness == Brightness.dark;

    final bgColor = widget.row.isChecked
        ? (isDark ? AppColors.darkGreenLight : AppColors.greenLight)
        : (isDark ? AppColors.darkSurface : AppColors.white);

    return GymCard(
      variant: widget.row.isChecked
          ? GymCardVariant.flat
          : GymCardVariant.outlined,
      backgroundColor: bgColor,
      padding: const EdgeInsets.all(16.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              // Exercise Number/Checkmark Badge
              Container(
                width: 36,
                height: 36,
                decoration: BoxDecoration(
                  color: widget.row.isChecked
                      ? AppColors.successGreen
                      : AppColors.energyOrange,
                  shape: BoxShape.circle,
                ),
                alignment: Alignment.center,
                child: Text(
                  widget.row.isChecked ? "✓" : "${widget.row.orderIndex + 1}",
                  style: GymTypography.titleMedium.white.bold,
                ),
              ),
              const SizedBox(width: 12),

              // Title and Focus Details
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Row(
                      children: [
                        Expanded(
                          child: Text(
                            widget.row.nameVi,
                            style: GymTypography.titleMedium.bold.copyWith(
                              color: customColors.primaryText,
                              decoration: widget.row.isChecked
                                  ? TextDecoration.lineThrough
                                  : TextDecoration.none,
                            ),
                          ),
                        ),
                        if (widget.row.isLightWorkout) ...[
                          const SizedBox(width: 6),
                          Container(
                            decoration: BoxDecoration(
                              color: AppColors.energyOrange.withValues(
                                alpha: 0.15,
                              ),
                              borderRadius: BorderRadius.circular(4),
                            ),
                            padding: const EdgeInsets.symmetric(
                              horizontal: 6,
                              vertical: 2,
                            ),
                            child: Text(
                              "Tập nhẹ",
                              style: GymTypography.labelSmall.orange.bold,
                            ),
                          ),
                        ],
                      ],
                    ),
                    const SizedBox(height: 2),
                    Row(
                      children: [
                        Text(
                          _muscleEmoji(widget.row.primaryMuscleGroup),
                          style: const TextStyle(fontSize: 14),
                        ),
                        const SizedBox(width: 4),
                        Expanded(
                          child: Text(
                            "${widget.row.prescriptionText} · nghỉ ${widget.row.restSeconds}s",
                            style: isDark
                                ? GymTypography.bodySmall.mutedDark
                                : GymTypography.bodySmall.muted,
                            overflow: TextOverflow.ellipsis,
                          ),
                        ),
                      ],
                    ),
                  ],
                ),
              ),

              // Circular Checkbox Toggle
              GymCheckbox(
                value: widget.row.isChecked,
                onChanged: widget.enabled
                    ? (val) => widget.onCheckedChange(val)
                    : null,
              ),
            ],
          ),

          // Instructions Trigger Link
          InkWell(
            onTap: widget.enabled
                ? () {
                    setState(() {
                      _expanded = !_expanded;
                    });
                  }
                : null,
            borderRadius: BorderRadius.circular(8),
            child: Padding(
              padding: const EdgeInsets.symmetric(vertical: 8.0),
              child: Row(
                children: [
                  Text(
                    _expanded ? "Ẩn hướng dẫn ▲" : "Xem hướng dẫn ▼",
                    style: GymTypography.labelSmall.orange.bold,
                  ),
                ],
              ),
            ),
          ),

          // Collapsible Instruction Sheet
          AnimatedSize(
            duration: const Duration(milliseconds: 250),
            curve: Curves.easeInOut,
            child: _expanded
                ? Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      const SizedBox(height: 8),
                      Container(
                        width: double.infinity,
                        decoration: BoxDecoration(
                          color: isDark
                              ? AppColors.darkSurfaceVariant
                              : AppColors.surfaceGray,
                          borderRadius: GymRadius.mdBorder,
                        ),
                        padding: const EdgeInsets.all(14.0),
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            // Step-by-Step Instructions
                            ...List.generate(widget.row.instructionsVi.length, (
                              index,
                            ) {
                              return Padding(
                                padding: const EdgeInsets.only(bottom: 6.0),
                                child: Row(
                                  crossAxisAlignment: CrossAxisAlignment.start,
                                  children: [
                                    Text(
                                      "${index + 1}.",
                                      style:
                                          GymTypography.bodyMedium.orange.bold,
                                    ),
                                    const SizedBox(width: 8),
                                    Expanded(
                                      child: Text(
                                        widget.row.instructionsVi[index],
                                        style: GymTypography.bodyMedium
                                            .copyWith(
                                              color: customColors.primaryText,
                                              height: 1.4,
                                            ),
                                      ),
                                    ),
                                  ],
                                ),
                              );
                            }),

                            if (widget.row.animationId != null) ...[
                              GymGap.md,
                              SizedBox(
                                width: double.infinity,
                                height: 48,
                                child: ElevatedButton(
                                  onPressed: () {
                                    Exercise3DDialog.show(
                                      context: context,
                                      animationId: widget.row.animationId!,
                                      exerciseName: widget.row.nameVi,
                                      instructions: widget.row.instructionsVi,
                                    );
                                  },
                                  style: ElevatedButton.styleFrom(
                                    backgroundColor: AppColors.energyOrange,
                                    foregroundColor: Colors.white,
                                    elevation: 0,
                                    shape: RoundedRectangleBorder(
                                      borderRadius: BorderRadius.circular(10),
                                    ),
                                    padding: EdgeInsets.zero,
                                  ),
                                  child: Text(
                                    "Xem chuyển động 3D",
                                    style: GymTypography.titleSmall.white.bold,
                                  ),
                                ),
                              ),
                            ],
                          ],
                        ),
                      ),
                    ],
                  )
                : const SizedBox.shrink(),
          ),

          // Exercise Substitution Triggers
          if (!widget.row.isChecked) ...[
            const SizedBox(height: 4),
            Align(
              alignment: Alignment.centerRight,
              child: TextButton(
                onPressed: widget.enabled ? widget.onSubstitute : null,
                style: TextButton.styleFrom(
                  padding: const EdgeInsets.symmetric(
                    horizontal: 12,
                    vertical: 4,
                  ),
                  minimumSize: Size.zero,
                  tapTargetSize: MaterialTapTargetSize.shrinkWrap,
                ),
                child: Text(
                  "Thay bài",
                  style: GymTypography.labelSmall.orange.bold,
                ),
              ),
            ),
          ],
        ],
      ),
    );
  }

  @override
  void didUpdateWidget(covariant ExerciseCard oldWidget) {
    super.didUpdateWidget(oldWidget);
    // Reset expand state if workout session or exercise ID changed
    if (oldWidget.row.exerciseId != widget.row.exerciseId ||
        oldWidget.sessionId != widget.sessionId) {
      setState(() {
        _expanded = false;
      });
    }
  }
}
